# DSCI 504

## Week 3

### Assignment: Selecting OPC Data for Review (Basic Queries)

### Student Name: SEIF KUNGULIO

<u>How to complete this assignment:</u> This Notebook has two types of cells, text, and code. Cell marked with **_Do not mark in this cell_** are not to be edited. Other text cells that require answers or typing may be edited by double-clicking on the cell and editing. Clicking off of the cell saves the text. Code cells are where you will enter in your SQL code. The results of your query will be displayed immediately beneath your code cell and are interactive. For full credit, you must write, execute, and save your code in your Notebook. The last cell of the Notebook has instructions for saving the Notebook as an html file for submission.

**_Do not mark in this cell_**

### Assignment Objective: At the end of this assignment, you will be able to conduct queries on single tables within the OPC database, to include unary table self-joins.

Scenario: You have been tasked with doing a data quality check for customers who share the same city... 

### 3.1 Question: Write a query using **only** the _customers_ table in the OPC database that returns **all** customers that share the **same** _customer city_. To qualify, <u>a city must be shared by more than one customer.</u>

#### Note: To answer this question you may need to use a self-join on a single table. We cover joins in Week 5, so I am including the wireframe for a self-join below for you to use as a reference.

<code>SELECT t1.col1 AS "Name 1",  
t1.col2 as "Name 2",`
t1.col as "Another Column"
FROM table t1  
JOIN table t2  
ON t1.col = t2.col  
AND t1.id <> t2.id  
ORDER BY t1.col1, t1.col2;</code>

In [1]:
SELECT 
    CONCAT(c1.cus_first_name, ' ', c1.cus_last_name) AS "Customer 1",
    CONCAT(c2.cus_first_name, ' ', c2.cus_last_name) AS "Customer 2",
    c1.cus_city AS "Shared City"
FROM 
    dsci_504.customers c1
JOIN 
    dsci_504.customers c2
    ON c1.cus_city = c2.cus_city
    AND c1.cus_id <> c2.cus_id
ORDER BY 
    c1.cus_city, c1.cus_first_name, c2.cus_first_name;

(135596 row(s) affected)

Displaying Top 5000 rows.

Total execution time: 00:00:05.624

Customer 1 Customer 2 Shared City Adam Aaronson Adam Williamson Arispe Adam Williamson Adam French Arispe Adam French Adam Aaronson Arispe Adam Aaronson Adam French Arispe Adam French Adam Williamson Arispe Adam Williamson Adam Aaronson Arispe Adam Williamson Airn Conley Arispe Adam French Airn Conley Arispe Adam Aaronson Airn Conley Arispe Adam Aaronson Albert Gorman Arispe Adam French Albert Gorman Arispe Adam Williamson Albert Gorman Arispe Adam Aaronson Brandon Etzler Arispe Adam Williamson Brandon Etzler Arispe Adam French Brandon Etzler Arispe Adam Williamson Brandt Maurer Arispe Adam French Brandt Maurer Arispe Adam Aaronson Brandt Maurer Arispe Adam French Brian Shelton Arispe Adam Aaronson Brian Shelton Arispe Adam Williamson Brian Shelton Arispe Adam French Brice Smith-Warren Arispe Adam Aaronson Brice Smith-Warren Arispe Adam Williamson Brice Smith-Warren Arispe Adam Williamson Carl Calahan Arispe Adam French Carl Calahan Arispe Adam Williamson Carl Harper Arispe Adam Aaronson Carl Calahan Arispe Adam French Carl Harper Arispe Adam Aaronson Carl Harper Arispe Adam Williamson Cooper Ramirez Arispe Adam French Cooper Ramirez Arispe Adam Aaronson Cooper Ramirez Arispe Adam Williamson Corey Conley Arispe Adam French Corey Conley Arispe Adam Aaronson Corey Conley Arispe Adam Aaronson Eric Delapena Arispe Adam French Eric Delapena Arispe Adam Williamson Eric Delapena Arispe Adam Williamson Erin Ridgley Arispe Adam French Erin Ridgley Arispe Adam Aaronson Erin Ridgley Arispe Adam Williamson Farris Moore Arispe Adam Aaronson Farris Roberts Arispe Adam Williamson Farris Roberts Arispe Adam French Farris Moore Arispe Adam Aaronson Farris Moore Arispe Adam French Farris Roberts Arispe Adam French George Ramirez Arispe Adam Williamson George Ramirez Arispe Adam Aaronson George Ramirez Arispe Adam Williamson Jack Dawson Arispe Adam Aaronson Jack Dawson Arispe Adam Aaronson Jack Tompsen Arispe Adam Williamson Jack Tompsen Arispe Adam French Jack Tompsen Arispe Adam French Jack Dawson Arispe Adam French Jasmine Calahan Arispe Adam Williamson Jasmine Calahan Arispe Adam Aaronson Jasmine Calahan Arispe Adam Aaronson Jonathon Jamieson Arispe Adam French Jonathon Jamieson Arispe Adam Williamson Jonathon Jamieson Arispe Adam Williamson Joshua Harper Arispe Adam Aaronson Joshua Harper Arispe Adam French Joshua Jamieson Arispe Adam Williamson Joshua Jamieson Arispe Adam French Joshua Harper Arispe Adam Aaronson Joshua Jamieson Arispe Adam French Leah Willis Arispe Adam Williamson Leah Willis Arispe Adam Aaronson Leah Willis Arispe Adam Aaronson Mads Cooper Arispe Adam French Mads Cooper Arispe Adam Williamson Mads Jackson Arispe Adam Aaronson Mads Jackson Arispe Adam Williamson Mads Cooper Arispe Adam French Mads Jackson Arispe Adam Williamson Maggs Jones Arispe Adam Aaronson Maggs Jones Arispe Adam Williamson Maggs Delacruz Arispe Adam French Maggs Jones Arispe Adam French Maggs Delacruz Arispe Adam Aaronson Maggs Delacruz Arispe Adam Aaronson Margo Jones Arispe Adam French Margo Jones Arispe Adam French Margo Seiler Arispe Adam Williamson Margo Jones Arispe Adam Williamson Margo Seiler Arispe Adam Aaronson Margo Seiler Arispe Adam Williamson Miguel Smith Arispe Adam Aaronson Miguel Smith Arispe Adam French Miguel Smith Arispe Adam Aaronson Mike Barrister Arispe Adam Williamson Mike Barrister Arispe Adam French Mike Barrister Arispe Adam Williamson Mindy Williamson Arispe Adam French Mindy Applewood Arispe Adam French Mindy Williamson Arispe Adam Williamson Mindy Applewood Arispe Adam Aaronson Mindy Applewood Arispe Adam Aaronson Mindy Williamson Arispe Adam French Nadine Rowley Arispe Adam French Nadine Manning Arispe Adam Aaronson Nadine Rowley Arispe Adam Williamson Nadine Rowley Arispe Adam Williamson Nadine Manning Arispe Adam Aaronson Nadine Manning Arispe Adam Williamson Noel Slain Arispe Adam Aaronson Noel Willis Arispe Adam French Noel Willis Arispe Adam French Noel Dewey Arispe Adam French Noel Slain Arispe Adam Aaronson Noel 

Scenario: You have been tasked with identifying products that may have the exact same product price...

### 3.2 Question: Write a query that returns <b>all</b> <em>products</em> that share the <b>same</b> <em>product price</em>.

#### Note: You may need to use the self-query template from above to assist in answering this question

In [2]:
-- Identify pairs of products from the same table that share the exact same price, using a self-join.
SELECT 
    p1.prod_name AS "Product 1",
    p2.prod_name AS "Product 2",
    p1.prod_price AS "Shared Price"
FROM dsci_504.products p1
JOIN dsci_504.products p2
    ON p1.prod_price = p2.prod_price
    AND p1.prod_id <> p2.prod_id
ORDER BY p1.prod_price, p1.prod_name, p2.prod_name;

(32 row(s) affected)

Total execution time: 00:00:00.056

Product 1 Product 2 Shared Price OIZ Primer 3799.00 Primer OIZ 3799.00 Clash Habit 3999.00 Clash Instinct 3999.00 Clash Occam 3999.00 Clash Occam 3999.00 Habit Clash 3999.00 Habit Instinct 3999.00 Habit Occam 3999.00 Habit Occam 3999.00 Instinct Clash 3999.00 Instinct Habit 3999.00 Instinct Occam 3999.00 Instinct Occam 3999.00 Occam Clash 3999.00 Occam Clash 3999.00 Occam Habit 3999.00 Occam Habit 3999.00 Occam Instinct 3999.00 Occam Instinct 3999.00 Occam Occam 3999.00 Occam Occam 3999.00 Foxy Ripley 4199.00 Ripley Foxy 4199.00 Insurgent Ripmo 4399.00 Ripmo Insurgent 4399.00 Shredd Dog Trail Pistol 5375.99 Trail Pistol Shredd Dog 5375.99 Altitude Firebird 5899.00 Firebird Altitude 5899.00 Rail Wreckoning 7399.00 Wreckoning Rail 7399.00

Scenario: You have been finding orders that do not have a ship date...

### 3.3 Question: Write a query that returns <b>all</b> <em>orders</em> where the <em>shipping date</em> is <b>NULL</b>.

In [3]:
SELECT *
FROM dsci_504.orders
WHERE ord_ship_date IS NULL;

(0 row(s) affected)

Total execution time: 00:00:00.051

ord_id ord_date ord_tax_loc order_tot ord_ship_add ord_ship_date ord_track_num warehouse_id cus_id

Scenario: You have been tasked with identifying if any orders have tracking numbers that appear in more than one order...

### 3.4 Question: Write a query that returns the count of <em>order tracking numbers</em> where the <em>order tracking number</em> has a count <b>Greater than 1</b>.

In [4]:
SELECT ord_track_num, COUNT(*) AS tracking_count
FROM dsci_504.orders
GROUP BY ord_track_num
HAVING COUNT(*) > 1;

(0 row(s) affected)

Total execution time: 00:00:00.127

ord_track_num tracking_count

Scenario: You have been finding build names that may sell well due to association with the movies Star Wars and Star Trek...

### 3.5 Question: Write a query that returns <b>all</b> <em>build names</em> where the <em>build name</em> contains the word <b>star</b>.

In [5]:
SELECT build_name
FROM dsci_504.builds
WHERE build_name LIKE '%star%';

(4 row(s) affected)

Total execution time: 00:00:00.063

build_name five star 3-star 2-star 1-star

Scenario: You have been tasked with finding our how many hardtail products OPC currently produces...

### 3.6 Question: Write a query that returns the count of <b>all</b> <em>products</em> where the <em>product description</em> is <b>Hardtail</b>. Do not use a SELECT *. Choose two meaningful columns from the <em>products</em> table to display.

In [6]:
SELECT prod_name, prod_description, COUNT(prod_name) AS hardtail_count
FROM dsci_504.products
WHERE prod_description LIKE '%Hardtail%'
GROUP BY prod_name, prod_description;

(4 row(s) affected)

Total execution time: 00:00:00.032

prod_name prod_description hardtail_count Air9 Hardtail Mountain Bike 1 Chameleon Hardtail Mountain Bike 1 DV9 Hardtail Mountain Bike 1 Scout Hardtail Mountain Bike 1

Scenario: You have tasked with finding out how many customers are from each state...

### 3.7 Question: Write a query that returns the <em>count</em> of <b>all</b> <em>customers</em> from each <em>state</em> in the <em>customers</em> table. Be sure to order by the state with the highest number of customers first.

In [7]:
SELECT state_id, COUNT(*) AS customer_count
FROM dsci_504.customers
GROUP BY state_id
ORDER BY customer_count DESC;

(30 row(s) affected)

Total execution time: 00:00:00.164

state_id customer_count 26 103 16 99 14 97 29 95 28 93 10 93 7 93 6 91 23 91 24 91 19 91 22 91 15 90 21 90 9 89 17 88 20 88 3 87 12 87 13 87 27 87 30 83 1 83 5 81 8 80 11 76 18 76 25 73 4 73 2 73

Scenario: You want to know what state has the highest customer count...

### 3.7a Question: Write a query that returns the <em>state</em> with the <em>id</em> from the query above.

In [8]:
SELECT c.state_id, s.state, COUNT(*) AS customer_count
FROM dsci_504.customers c
JOIN dsci_504.states s ON c.state_id = s.state_id
GROUP BY c.state_id, s.state
ORDER BY customer_count DESC;

(30 row(s) affected)

Total execution time: 00:00:00.130

state_id state customer_count 26 UT 103 16 ME 99 14 KY 97 29 WV 95 10 HI 93 7 DE 93 28 WA 93 19 NE 91 23 RI 91 6 CT 91 22 PA 91 24 SC 91 15 LA 90 21 OR 90 9 GA 89 20 OH 88 17 MN 88 3 AZ 87 12 IN 87 13 KS 87 27 VA 87 1 AK 83 30 WY 83 5 CO 81 8 FL 80 11 IA 76 18 NC 76 4 CA 73 2 AL 73 25 TN 73

Scenario: You have been tasked with finding all orders placed in a specific date range...

### 3.8 Question: Write a query that returns <b>all</b> <em>orders</em> placed in <b>March 2019</b>.

In [9]:
SELECT *
FROM dsci_504.orders
WHERE ord_date >= '2019-03-01'
  AND ord_date < '2019-04-01';

(7 row(s) affected)

Total execution time: 00:00:00.074

ord_id ord_date ord_tax_loc order_tot ord_ship_add ord_ship_date ord_track_num warehouse_id cus_id 4066723 2019-03-28 21 4399.00 False 2019-03-29 PAC9045159-1 2 48 2344343 2019-03-09 17 4701.43 True 2019-03-12 PAC8075556-0 3 9 4733459 2019-03-21 29 3189.54 False 2019-03-29 PAC8887258-0 1 2140 3015647 2019-03-17 13 6175.94 False 2019-03-21 PAC4349549-3 2 710 2747613 2019-03-07 30 3950.96 False 2019-03-17 PAC820955-3 3 713 2716051 2019-03-13 12 4523.95 True 2019-03-20 PAC546684-2 2 1081 3496466 2019-03-28 4 4717.93 False 2019-04-06 PAC3061964-3 3 2183

Scenario: You have been tasked with finding all customers whom joined in a specific date range...

### 3.9 Question: Write a query that returns <b>all</b> <em>customers</em> whom joined between <b>Sept 1 2001</b> and <b>Sept 30 2001</b>.

In [10]:
SELECT *
FROM dsci_504.customers
WHERE cus_join_date >= '2001-09-01'
  AND cus_join_date <= '2001-09-30';

(15 row(s) affected)

Total execution time: 00:00:00.081

cus_id cus_num cus_last_name cus_first_name cus_add_num cus_address cus_city state_id cus_zip cus_phone cus_join_date cus_app_cd cus_app_num 356 Del73426 Delacruz Margo 63740 Crystal Spring Corners Mcarthur Estates 2 89443 3875686732 2001-09-25 6 OPC80718306 464 She39200 Sherman George 84889 Honey Zephyr Nook Winkelman 3 78430 1605936456 2001-09-30 1 OPC85270843 495 Del44118 Delacruz Brandon 53556 Apache Wolfs Mill 18 18151 7593410556 2001-09-23 1 OPC79747551 1121 Jim75660 Jiminez Summer 29615 Rocky Creek Meadow Klagetoh 8 41779 8952766 2001-09-06 1 OPC86250057 1189 Coo83584 Cooper Reginald 11391 Strom Laffertys Trailer Court 13 64429 1605847977 2001-09-01 5 OPC87744241 1587 App17645 Applewood Miguel 20572 Rimer Kwikpak 17 60788 6915803429 2001-09-24 4 OPC84950666 1907 Aco67531 Acosta Joshua 84384 Cinder Woods Norton Shores 29 82569 7256978201 2001-09-16 3 OPC82133096 1936 Man97461 Manchester Kenadi 15274 Colonial Panda Pike Tiptop 11 76588 4289060045 2001-09-30 6 OPC84539106 1998 Gar42269 Garth Richard 45839 Silver Lake Crossing Crystal Lake 19 25823 912849216 2001-09-09 2 OPC79316181 2141 Bro77695 Brooks Summer 29242 Stan Sleuter Fomer 19 54940 8105755047 2001-09-23 4 OPC84407421 2194 Moo61600 Moore Joshua 84080 Carbon Hill-Buchtel Wolfs Mill 28 5587 4523182148 2001-09-12 4 OPC81668931 2257 Moo82033 Moore Taylor 2297 Quaking Embers Quay Brownsburg 18 95851 7742904787 2001-09-04 5 OPC81423068 2330 San99486 Sanchez Erin 46434 Silver Dale Jetty Vega Park 6 2155 3442359879 2001-09-02 3 OPC79195861 2333 App89762 Applewood Timothy 12734 Jagged Branch Mews Crystal Lake 6 26523 179933424 2001-09-03 1 OPC86619531 2351 Tom53631 Tompsen Piper 19985 Umber Forest Arispe 28 309 9976832027 2001-09-04 3 OPC80023511

### 3.10 Write your own query on a single table that identifies a piece of infroamtion that peaks your interest.

In [11]:
SELECT ord_date, COUNT(*) AS total_orders
FROM dsci_504.orders
GROUP BY ord_date
ORDER BY total_orders DESC
LIMIT 5;

(5 row(s) affected)

Total execution time: 00:00:00.085

ord_date total_orders 2016-06-30 4 2005-11-10 4 2010-08-07 3 2010-06-14 3 2008-03-09 3

## Testing your Knowledge

Answer the following questions:

#### 1\. Do you think the self-join you created in 3.1 assists in answering any valuable questions about how many cities have more than one customer? Why might this not be an optimal query

The self-join query is effective for identifying pairs of customers who live in the same city. By joining the **customers** table to itself on the \`cus\_city\` field and excluding matches with identical customer IDs, it highlights customer pairs sharing the same location. This is useful for answering questions like, "Which customers are from the same city?" or "What customer pairs exist by city?"

<span style="color: var(--vscode-foreground);">However, this approach is not ideal for determining how many cities have more than one customer. The self-join generates multiple rows for each pair, which can lead to unnecessary complexity and potential overcounting. For instance, if three customers are from the same city, the query produces three pairings (A-B, A-C, B-C), complicating a straightforward count.</span>

<span style="color: var(--vscode-foreground);">A more efficient method is to use a <b>GROUP BY</b> query with a <b>HAVING</b> clause. This groups customers by city, counts how many are in each group, and filters out cities with only one customer. The result is a cleaner, more accurate, and scalable way to identify cities with multiple customers.</span>

<span style="color: var(--vscode-foreground);">In conclusion, while the self-join query is useful for analyzing relationships between customers in the same city, it is not well-suited for counting cities with more than one customer. For that purpose, aggregation using <b>GROUP BY</b>&nbsp;and <b>HAVING</b>&nbsp;provides a more effective solution.</span>

#### 2\. What product price was shared by the most products in 3.2? How would you recommend reviewing additiona information about these products?

The most commonly shared product price in the query result is **$3999.00**, associated with five products: **Clash**, **Habit**, **Instinct**, and **Occam**. To understand why these products share the same price, it's recommended to query additional details like category and description. Analyzing this data can reveal similarities in features or market positioning.

#### 3\. Did 3.3 return any data? Is it common for a query to return nothing? What benefit may asking this type of a question have for an analyst?

The query in 3.3 returned no data, meaning there are no orders with a missing shipping date. It's common for queries to return no results, and this can be useful for analysts to confirm data completeness and ensure there are no issues, such as unprocessed or delayed orders.

#### 4\. Where there any tracking numbers that appeared more than once in 3.4? Is this type of query useful in detecting duplicates? Did you go through any other tables and try to find duplicates? Why might you not find any in core tables?

The query in 3.4 found no duplicate tracking numbers, indicating all are unique. This type of query is useful for detecting duplicates and ensuring data integrity. The absence of duplicates in core tables may be due to good database design and enforced constraints that prevent such issues.

#### 5\. Go back and look at 3.9. You may need to run some additional queries to expand further into teh rabbit hole, but find two to four customers in the query results in 3.9 that may or may not be family. Is there anything unique about these customers? If so, what might it be?

Upon reviewing the results from query 3.9, which returned customers who joined between September 1 and September 30, 2001, several customer records suggest potential family relationships or notable patterns worth exploring further.

One possible family pair is **Margo Delacruz** (cus\_id 356) and **Brandon Delacruz** (cus\_id 495). They share the same last name and joined just two days apart—Margo on September 25 and Brandon on September 23. This close timing may indicate a coordinated signup, possibly influenced by a family referral or promotional offer.

Another potential family group includes **Joshua Moore** (cus\_id 2194) and **Taylor Moore** (cus\_id 2257). Although they list different cities—Wolfs Mill and Brownsburg—the shared last name and similar join dates (September 12 and September 4, respectively) suggest they might be related, such as siblings or other close relatives who signed up around the same time.

A third pair to consider is **Miguel Applewood** (cus\_id 1587) and **Timothy Applewood** (cus\_id 2333). The Applewood surname is fairly distinctive, and both customers joined in early September, which may also suggest a familial connection or shared household.

These observations are valuable from an analytical perspective. Detecting potential family ties can inform customer segmentation strategies, improve targeting for household-based promotions, and assist in tracking referrals. To confirm these relationships, additional queries comparing addresses, phone numbers, or application details would provide further insights.

## Assignment Takeaway

### **Summarize this week's assignment in your own words and identify what you think you learned? Was it valuable? What can be improved?**

This week’s assignment focused on querying single tables within the OPC database, with a particular emphasis on **self-joins**, **data validation**, and **exploratory analysis**. I learned how to use **self-joins** to identify relationships within a single table—such as customers from the same city or products that share the same price. I also practiced filtering for **NULL** values, detecting **duplicate entries**, and using **aggregate functions** with **GROUP BY** and **HAVING** to analyze data quality and patterns.

One valuable takeaway was understanding the difference between when to use a **self-join** versus when a **GROUP BY approach** is more efficient, such as in counting distinct categories or detecting unique values. The scenarios also highlighted the importance of verifying **data integrity**, like checking for missing shipping dates or repeated tracking numbers. I appreciated the opportunity to dig deeper into data relationships, such as exploring customer join dates to infer potential family ties.

Overall, the assignment was **very valuable** in reinforcing SQL skills and analytical thinking. It provided a realistic context for using queries not just to retrieve data, but to **validate, analyze, and draw insights**. If anything could be improved, it would be helpful to include **more guidance on interpreting results** or follow-up tasks that encourage turning those insights into business recommendations. Nonetheless, it was a strong, hands-on learning experience.

Copy the below code into a new Code cell and change the _Kernel_ above to _Python 3_. This will generate an html file of this Notebook in your current folder for submission. Submit the html file in Canvas.

#### Note: You may need to install the kernel. If you system already has one, ADS will find it. Click Next, then Install. The Output will display below. Once done, you will see "Jupyter is running at [http://localhost:8888/..."](http://localhost:8888/...&quot;) This only needs to be done once.

#### When you run the below code, if you see:

\[NbConvertApp\] Converting notebook DSCI504\_Wk2\_Ass1\_SimpleQueries.ipynb to html  
\[NbConvertApp\] Writing 570001 bytes to DSCI504\_Wk2\_Ass1\_SimpleQueries.html

#### then you successfully saved your file to html and it will be in the current folder where your assignment file is located.

`!jupyter nbconvert --to html DSCI504_Wk2_Ass1_SimpleQueries_Key.ipynb`

In [1]:
!jupyter nbconvert --to html DSCI504_Wk3_Assignment.ipynb

[NbConvertApp] Converting notebook DSCI504_Wk3_Assignment.ipynb to html


[NbConvertApp] Writing 686198 bytes to DSCI504_Wk3_Assignment.html
